### PR TITLE
Refactor Popover

### PR DESCRIPTION
**Pull request by LennertBlommaert**
### Description

This PR refactors the [`Popover` component](https://components.teamleader.design/?knob-Locale=nl&knob-Size=medium&knob-Color=neutral&knob-Tint=lightest&selectedKind=Popover&selectedStory=vertical&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs).

And solves the issue #303: it now renders more accurately, by not blindingly choosing the opposite `direction`, if it would be rendered off-screen in the given one. 
In the last case, it chooses for the direction with the most space and if needed, it becomes vertically scrollable.

### Screenshots

In this example, the `direction` is set to `'south'`.

_When `Popover` fits in viewport, following the given `direction`:_
screen shot 2018-09-03 at 18 01 25

_When `Popover` does not fit in the viewport, following both the given and its opposite `direction` (and there is more space in the current direction):_
screen shot 2018-09-03 at 18 01 37

_When `Popover` does not fit in the viewport, following both the given and its opposite `direction` (and there's more space in the opposite direction):_
screen shot 2018-09-03 at 18 01 51

### Breaking changes

None.

### CODE REVIEW

File: components/popover/positionCalculation.js [hunk 1]
```diff
@@ -1,40 +1,35 @@
...
+    middle: top + (bottom - top) / 2,
+    center: left + (right - left) / 2,
+  };
+};
 
-function getTargetPosition(targetEl) {
-  const targetRect = targetEl.getBoundingClientRect();
-  const width = Math.round(targetRect.width);
-  const height = Math.round(targetRect.height);
+const getTargetPositionValues = targetEl => {
```
Comment (driesd) [inline components/popover/positionCalculation.js]: PP: getPopoverPositionValues

File: components/popover/positionCalculation.js [hunk 2]
```diff
@@ -92,37 +79,75 @@ const updateHorizontalPositionIfNeeded = (position, anchorPosition, targetPositi
...
-  arrowLeft: 2 * POPUP_OFFSET,
-});
-
-const positionRight = (anchorPosition, targetPosition, offsetCorrection) => ({
-  left: anchorPosition.right - targetPosition.width + offsetCorrection,
-  arrowLeft: targetPosition.width - POPUP_OFFSET * 3,
-});
+const getPositionTopValue = ({ direction, anchorPosition, targetPosition }) =>
+  direction === 'north'
+    ? anchorPosition.Top - targetPosition.height - POPUP_OFFSET
```
Comment (driesd) [inline components/popover/positionCalculation.js]: Typo: anchorPosition`.Top` > anchorPosition`.top`

File: components/popover/positionCalculation.js [hunk 3]
```diff
@@ -92,37 +79,75 @@ const updateHorizontalPositionIfNeeded = (position, anchorPosition, targetPositi
...
+    case 'right':
+      return targetPosition.width - POPUP_OFFSET * 3;
+  }
+};
 
-const directionSouth = anchorPosition => ({
-  top: anchorPosition.top + anchorPosition.height + POPUP_OFFSET,
-  arrowTop: -ARROW_OFFSET,
-});
+const getDirectionRendersOnScreen = ({ direction, anchorPosition, targetPosition }) => {
```
Comment (driesd) [inline components/popover/positionCalculation.js]: PP: const `isInViewport`

File: components/popover/positionCalculation.js [hunk 4]
```diff
@@ -92,37 +79,75 @@ const updateHorizontalPositionIfNeeded = (position, anchorPosition, targetPositi
...
+      return 'north';
+    case 'east':
+      return 'west';
+    case 'west':
+      return 'east';
   }
+};
 
-  return anchorPosition.bottom + targetPosition.height + POPUP_OFFSET > window.innerHeight ? 'north' : 'south';
+const getOppositeDirectionRendersOnScreen = ({ direction, anchorPosition, targetPosition }) =>
```
Comment (driesd) [inline components/popover/positionCalculation.js]: Maybe you can eliminate this function and use `isInViewport` with a different `direction` parameter?

**Comment by nickwaelkens** [comment]
Amazing work so far @LennertBlommaert. I think you can go a step further and get rid of the `factory` pattern here, quite useless if you check what is being passed through and how it's used. Hit me up if you need any help!

File: components/popover/Popover.js [hunk 5]
```diff
@@ -7,36 +7,23 @@ import InjectOverlay from '../overlay';
...
 
-    state = {
-      positioning: {
-        left: 0,
-        top: 0,
-        arrowLeft: 0,
-        arrowTop: 0,
-      },
-    };
+    state = { positioning: { left: 0, top: 0, arrowLeft: 0, arrowTop: 0, maxPopoverHeight: 'inital' } };
```
Comment (nickwaelkens) [inline components/popover/Popover.js]: Typo: `initial`?

**Pull request merged**